### ISSUE TITLE
Explicit type arguments are required to avoid warnings using generic TryGetValue method

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#48992]
Warnings are reported for the first two uses of `TryGetValue()`. To avoid warnings, it's necessary to use an explicit type argument.

```C#
#nullable enable

using System.Diagnostics.CodeAnalysis;

class Collection
{
    public bool TryGetValue<T>(object key, [MaybeNullWhen(false)] out T value)
    {
        value = default;
        return false;
    }
}
    
class Program
{
    static string GetValue(Collection c, object key)
    {
        // out string
        if (c.TryGetValue(key, out string s1)) // warning CS8600: Converting possible null value to non-nullable
        {
            return s1;
        }
        
        // out string?
        if (c.TryGetValue(key, out string? s2))
        {
            return s2; // warning CS8603: Possible null reference return
        }
        
        // out string?, explicit type argument
        if (c.TryGetValue<string>(key, out string? s3))
        {
            return s3;
        }
               
        // out var, explicit type argument
        if (c.TryGetValue<string>(key, out var s4))
        {
            return s4;
        }

        return string.Empty;
    }
}
```

From discussion with @RikkiGibson.

**Comment by RikkiGibson** [comment]
After discussion we are considering solving it with the following rule:

When analyzing an argument, if the original definition's corresponding parameter is a type parameter with conditional attributes (MaybeNullWhen, etc.), we suppress any W-warning that would have been produced by the postcondition.

Thus we can still infer the correct type argument (i.e. `string` in the above example) without introducing a difficult-to-suppress warning. Note that the language of this rule may need to be adjusted to make sure it addresses the particular scenarios we want without removing warnings from scenarios that we didn't want to change.

**Comment by cston** [comment]
From a compiler team discussion, the calls to `TryGetValue<T>()` with implicit type arguments seem ambiguous.

Did the caller of `c.TryGetValue(key, out string s1)` expect a non-`null` `string` associated with `key` or might the collection contain a `null` `string` value but the local annotation is missing? Did the caller of `c.TryGetValue(key, out string? s2)` expect a nullable `string` associated with `key` or was the local annotation for the failure case? The warnings in those two implicit cases encourage the caller to make the expected nullability explicit.

This is a narrow scenario since it involves a generic `TryGetValue<T>()` method rather than a non-generic method on a generic container type (such as `Dictionary<TKey, TValue>.TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)`).

Given the narrow scenario and the ambiguities when inferring nullability of type arguments, we will leave the behavior as-is.
